Set CMAKE_SYSTEM_NAME when cross-compiling

tmiasko opened the pull request:
Configure CMAKE_SYSTEM_NAME when cross-compiling in `configure_cmake`,
to tell CMake about target system. Previously this was done only for
LLVM step and now applies more generally to steps using cmake.

Helps with #74576.

rust-highfive commented:
r? @Mark-Simulacrum

(rust_highfive has picked a reviewer for you, use r? to override)

tmiasko committed:
Set CMAKE_SYSTEM_NAME when cross-compiling

Configure CMAKE_SYSTEM_NAME when cross-compiling in `configure_cmake`,
to tell CMake about target system. Previously this was done only for
LLVM step and now applies more generally to steps using cmake.

tmiasko commented:
@Mark-Simulacrum I would appreciate the review of this, to unblock #74576, where incorrect evaluation of CMAKE_SYSTEM_NAME breaks the tsan build. Thanks.

Mark-Simulacrum commented:
https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_NAME.html seems to indicate that we should also be setting CMAKE_SYSTEM_VERSION - is that not the case?

tmiasko commented:
Yes, I would expect that should be done as well. 

Fixing this might be non-trivial since rustc targets do not include version.
After running a FreeBSD cross-build on CI there is no indication that it is
necessary and looking at LLVM CMakeFiles checks for system versions are mostly
limited to tests. So I would rather avoid attempts to add version in this PR.

Mark-Simulacrum commented:
Seems reasonable. r=me with a comment added about the version not being set and our belief that not setting it should be fine (basically your last comment inserted into the code).

tmiasko commented:
Added a comment about CMAKE_SYSTEM_VERSION.

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 91f87bc9bc29461513aaa6252905a7ac0ac303e9 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 91f87bc9bc29461513aaa6252905a7ac0ac303e9 -->
<!-- homu: {"type":"Approved","sha":"91f87bc9bc29461513aaa6252905a7ac0ac303e9","approver":"Mark-Simulacrum"} -->
